stl/inc/stop_token

```diff
@@ -0,0 +1,401 @@
+                                                 // can detach without blocking on the callback
+
+            _Head->_Fn(_Head); // might destroy *_Head
+        }
+    }
+};
+
+class stop_source;
+
+class stop_token {
+    friend stop_source;
+    friend _Stop_callback_base;
+
+public:
+    stop_token() noexcept : _State{} {}
+    stop_token(const stop_token& _Other) noexcept : _State{_Other._State} {
+        const auto _Local = _State;
+        if (_Local != nullptr) {
+            _Local->_Stop_tokens.fetch_add(1, memory_order_relaxed);
+        }
+    }
+
+    stop_token(stop_token&& _Other) noexcept : _State{_STD exchange(_Other._State, nullptr)} {}
+    stop_token& operator=(const stop_token& _Other) noexcept {
+        stop_token{_Other}.swap(*this);
+        return *this;
+    }
+
+    stop_token& operator=(stop_token&& _Other) noexcept {
+        stop_token{_STD move(_Other)}.swap(*this);
+        return *this;
+    }
+
+    ~stop_token() {
+        const auto _Local = _State;
+        if (_Local != nullptr) {
+            if (_Local->_Stop_tokens.fetch_sub(1, memory_order_acq_rel) == 1) {
+                delete _Local;
+            }
+        }
+    }
+
+    void swap(stop_token& _Other) noexcept {
+        _STD swap(_State, _Other._State);
+    }
+
+    [[nodiscard]] bool stop_requested() const noexcept {
+        const auto _Local = _State;
+        return _Local != nullptr && _Local->_Stop_requested();
+    }
+
+    [[nodiscard]] bool stop_possible() const noexcept {
+        const auto _Local = _State;
+        return _Local != nullptr && _Local->_Stop_possible();
+    }
+
+    [[nodiscard]] friend bool operator==(const stop_token& _Lhs, const stop_token& _Rhs) noexcept {
+        return _Lhs._State == _Rhs._State;
+    }
+
+    friend void swap(stop_token& _Lhs, stop_token& _Rhs) noexcept {
+        _STD swap(_Lhs._State, _Rhs._State);
+    }
+
+private:
+    explicit stop_token(_Stop_state* const _State_) : _State{_State_} {}
+
+    _Stop_state* _State;
+};
+
+class stop_source {
+public:
+    stop_source() : _State{new _Stop_state} {}
+    explicit stop_source(nostopstate_t) noexcept : _State{} {}
+    stop_source(const stop_source& _Other) noexcept : _State{_Other._State} {
+        const auto _Local = _State;
+        if (_Local != nullptr) {
+            _Local->_Stop_sources.fetch_add(2, memory_order_relaxed);
+        }
+    }
+
+    stop_source(stop_source&& _Other) noexcept : _State{_STD exchange(_Other._State, nullptr)} {}
+    stop_source& operator=(const stop_source& _Other) noexcept {
+        stop_source{_Other}.swap(*this);
+        return *this;
+    }
+
+    stop_source& operator=(stop_source&& _Other) noexcept {
+        stop_source{_STD move(_Other)}.swap(*this);
+        return *this;
+    }
+
+    ~stop_source() {
+        const auto _Local = _State;
+        if (_Local != nullptr) {
+            if ((_Local->_Stop_sources.fetch_sub(2, memory_order_relaxed) >> 1) == 1) {

```

Comment:
In `barrier` ( #1057 ), I used named things:
```C++
inline constexpr ptrdiff_t _Barrier_arrival_token_mask = 1;
inline constexpr ptrdiff_t _Barrier_value_mask         = ~_Barrier_arrival_token_mask;
inline constexpr ptrdiff_t _Barrier_value_shift        = 1;
inline constexpr ptrdiff_t _Barrier_invalid_token      = 0;
inline constexpr ptrdiff_t _Barrier_value_step         = 1 << _Barrier_value_shift;
inline constexpr ptrdiff_t _Barrier_max                = (1ULL << (sizeof(ptrdiff_t) * CHAR_BIT - 2)) - 1;
```

I suggest that either you add named constants or I remove them.

Reply:
With named things it can go without shift, though anyway more verbose:
```C++
if ((_Local->_Stop_sources.fetch_sub(_Stop_source_count_step, memory_order_relaxed) == _Stop_source_count_step)
```

Reply:
Name all the things. Ask Kvothe

Reply:
I don't have great ideas for names here. Do you have suggestions?

Reply:
```
(fetch_sub(_Stop_sources_count_step, mo_relaxed)  >> _Stop_sources_count_shift) == 1
```
or
```
(fetch_sub(_Stop_sources_count_step`, mo_relaxed)  >> _Stop_sources_count_mask) == _Stop_sources_count_step
```

Reply:
I'm usually the arguer for long names and naming more things but I don't think they actually clarify what's going on here: the reader still needs to be intimately familiar with the bit-packing we're doing here. That is to say, a reader can't determine that the line is correct with the names but not determine that it's correct without the names.